File: kotlin/src/test/kotlin/jsontag/TestTypes.kt
package jsontag

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
sealed class Shape

@Serializable
@SerialName("Circle")
data class Circle(val radius: Double) : Shape()

@Serializable
@SerialName("Rectangle")
data class Rectangle(val width: Double, val height: Double) : Shape()

@Serializable
sealed class Vehicle

@Serializable
@SerialName("Car")
data class PersonalVehicle(val make: String) : Vehicle()

@Serializable
@SerialName("Truck")
data class TransportationVehicle(val payloadTons: Double) : Vehicle()

File: scala/src/test/scala/jsontag/TestTypes.scala
package jsontag

sealed trait Shape derives JsonTag.ReadWriter
case class Circle(radius: Double) extends Shape
case class Rectangle(width: Double, height: Double) extends Shape

sealed trait Vehicle derives JsonTag.ReadWriter

@upickle.implicits.key("Car")
case class PersonalVehicle(make: String) extends Vehicle

@upickle.implicits.key("Truck")
case class TransportationVehicle(payloadTons: Double) extends Vehicle
